common/poller/BasePoller.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (C) 2021 Vaticle
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.vaticle.typedb.core.common.poller;
+
+import com.vaticle.typedb.core.common.iterator.FunctionalIterator;
+
+import java.util.Optional;
+
+public class BasePoller<T> extends AbstractPoller<T> {
+
+    private final FunctionalIterator<T> source;
+
+    public BasePoller(FunctionalIterator<T> iterator) {
+        source = iterator;
+    }
+
+    @Override
+    public Optional<T> poll() {
+        if (source.hasNext()) return Optional.of(source.next());
+        source.recycle(); // TODO: Can we call hasNext() after recycle()?
```

Comment:
not sure we want to call `recycle()` eagerly here yeah, it should only be recycled when the BasePoller itself is recycled

Reply:
I also think it's more in our style to write this kinda thing as explicit `if-else` blocks even though both return (haikal's argument is it makes the flow easier to read)